### PR TITLE
os/driver/pm: Add pm driver for lock and unlock

### PR DESCRIPTION
**Pull request by ritesh55555**
this commit adds pm driver for locking and unlocking of specific PM state.

### CODE REVIEW

File: os/drivers/pm/pm.c [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+#include <debug.h>
+#include <assert.h>
+
+#define PM_DRVPATH           "/dev/pm"
+
+/****************************************************************************
+ * Function Prototypes
+ ****************************************************************************/
+
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
```
Comment (ewoodev) [inline os/drivers/pm/pm.c:37]: minor, please align order read, write ioctl

File: os/drivers/pm/pm.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
+    
+	/* Handle built-in ioctl commands */
+	switch (cmd) {
+    	case PMIOC_LOCK:
+            	if (arg >= PM_NORMAL && arg <= PM_SLEEP) {
+                	pm_stay(PM_IDLE_DOMAIN, arg);
```
Comment (ewoodev) [inline os/drivers/pm/pm.c:96]: please change domain to argument

File: os/drivers/pm/pm.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+ *   This api can be used to perform PM operation.
+ *
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
+    
+	/* Handle built-in ioctl commands */
+	switch (cmd) {
+    	case PMIOC_LOCK:
```
Comment (ewoodev) [inline os/drivers/pm/pm.c:94]: Please rename suspend/resume

File: os/include/tinyara/pm/pm.h [hunk 4]
```diff
@@ -313,6 +307,12 @@ enum pm_timer_type_e {
 	/* Scope for future expansion, up to 8 timer types can be supported */
 };
 
+/* This structure is used to send data to pm driver from app side */
+struct pm_header_s {
```
Comment (ewoodev) [inline os/include/tinyara/pm/pm.h:311]: why it is header??

how about argument or other word?

Comment (ritesh55555) [reply os/include/tinyara/pm/pm.h]: updated

**Review by ewoodev — APPROVED** [review]

File: os/drivers/Makefile [hunk 5]
```diff
@@ -88,6 +88,7 @@ include lwnl${DELIM}Make.defs
 include net$(DELIM)Make.defs
```
Comment (ewoodev) [inline os/drivers/Makefile:88]: Please change commit description pm lock unlock to suspend resume

Comment (ritesh55555) [reply os/drivers/Makefile:88]: updated

File: os/include/tinyara/pm/pm.h [hunk 6]
```diff
@@ -313,6 +307,12 @@ enum pm_timer_type_e {
 	/* Scope for future expansion, up to 8 timer types can be supported */
 };
 
+/* This structure is used to send data to pm driver from app side */
+struct pm_arg_s {
+	uint8_t pm_domain;
```
Comment (ewoodev) [inline os/include/tinyara/pm/pm.h:312]: Don't you need to limit this to enum?

Comment (ritesh55555) [reply os/include/tinyara/pm/pm.h]: done

**Review by ewoodev — APPROVED** [review]

File: os/include/tinyara/pm/pm.h [hunk 7]
```diff
@@ -313,6 +307,12 @@ enum pm_timer_type_e {
 	/* Scope for future expansion, up to 8 timer types can be supported */
 };
 
+/* This structure is used to send data to pm driver from app side */
+struct pm_arg_s {
+	enum pm_domain_e pm_domain;
```
Comment (ewoodev) [inline os/include/tinyara/pm/pm.h:312]: how about as below?
pm_arg.pm_domain  -> pm_arg.domain
pm_arg.pm_state -> pm_arg.state

File: os/include/tinyara/pm/pm.h [hunk 8]
```diff
@@ -313,6 +307,12 @@ enum pm_timer_type_e {
...
+	enum pm_state_e pm_state;
+};
+
```
Comment (ewoodev) [inline os/include/tinyara/pm/pm.h:315]: It will be used a lot from apps

how about `typedef pm_arg_t struct pm_arg_s` ?

Comment (ritesh55555) [reply os/include/tinyara/pm/pm.h:315]: done

File: os/drivers/pm/pm.c [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
+    
+	/* Handle built-in ioctl commands */
+	switch (cmd) {
+    	case PMIOC_SUSPEND:
+	        pm_stay(((struct pm_arg_s *)arg)->domain, ((struct pm_arg_s *)arg)->state);
```
Comment (ewoodev) [inline os/drivers/pm/pm.c:95]: please check pm_arg_t

File: os/board/rtl8730e/src/component/bluetooth/example/ble_scatternet/scatternet.c [hunk 10]
```diff
@@ -295,7 +295,7 @@ static rtk_bt_evt_cb_ret_t ble_tizenrt_scatternet_gap_app_callback(uint8_t evt_c
         if (!conn_ind->err) {
```
Comment (ewoodev) [inline os/board/rtl8730e/src/component/bluetooth/example/ble_scatternet/scatternet.c:295]: Pleaes check commit title

[This is a commit rule](https://github.com/Samsung/TizenRT/wiki/Documentations#commit-rules)

**Review by ewoodev — APPROVED** [review]

**Review by sunghan-chang — APPROVED** [review]

File: os/drivers/pm/pm.c [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+ *   for TUNEFREQ, arg should be an int type.
+ *
+ * Description:
+ *   This api can be used to perform PM operation.
+ *
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
+
```
Comment (Taejun-Kwon) [inline os/drivers/pm/pm.c:91]: you casting arg as (pm_art *), so null check seems necessary.

File: os/drivers/pm/pm.c [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+ *   for PMIOC_RESUME, arg is unsigned long representing PM STATE
+ *   for TUNEFREQ, arg should be an int type.
+ *
+ * Description:
+ *   This api can be used to perform PM operation.
+ *
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
```
Comment (Taejun-Kwon) [inline os/drivers/pm/pm.c:90]:
```
int ret = OK;
...
default :
      ret = - ENOSYS;

return ret;

**Pull request merged**